src/serialization/container.h

```diff
@@ -52,13 +52,32 @@ namespace serialization
       static constexpr const bool previously_varint = std::is_same<uint64_t, T>() || std::is_same<uint32_t, T>();
 
       if (!previously_varint && ar.varint_bug_backward_compatibility_enabled() && !typename Archive::is_saving())
-        return ::do_serialize(ar, e);
+        return do_serialize(ar, e);
       ar.serialize_varint(e);
       return true;
     }
 
+    //! @brief Add an element to a container, inserting at the back if applicable.
+    template <class Container>
+    auto do_add(Container &c, typename Container::value_type &&e) -> decltype(c.emplace_back(e), void())
+    { c.emplace_back(e); }
+    template <class Container>
+    auto do_add(Container &c, typename Container::value_type &&e) -> decltype(c.emplace(e), void())
+    { c.emplace(e); }
+
+    //! @brief Reserve space for N elements if applicable for container.
     template <typename C>
     void do_reserve(C &c, size_t N) {}
+    template <typename T> void do_reserve(std::vector<T> &c, size_t N) { c.reserve(N); }
```

Comment:
You can make the first function a template pack and then auto-return type reserve:
```c++
template<typename... C>
void do_reserve(const C&...) {}

template<typename C>
auto do_reserve(C& c, std::size_t N) -> decltype(c.reserve(N)) { return c.reserve(N); }
```
which gets all functions with a reserve.

Reply:
I like this! I was trying to find a not-horribly-verbose way to say this without mentioning any specific container